packages/sdk/src/peer_manager/peer_manager.ts

```diff
@@ -29,104 +72,186 @@ export class PeerManager {
+    this.libp2p.addEventListener(
+      "peer:connect",
+      this.onConnected as Libp2pEventHandler
+    );
+    this.libp2p.addEventListener(
+      "peer:disconnect",
+      this.onDisconnected as Libp2pEventHandler
+    );
   }
 
   public stop(): void {
-    this.stopConnectionListener();
+    this.libp2p.removeEventListener(
+      "peer:connect",
+      this.onConnected as Libp2pEventHandler
+    );
+    this.libp2p.removeEventListener(
+      "peer:disconnect",
+      this.onDisconnected as Libp2pEventHandler
+    );
   }
 
-  public getPeers(): PeerId[] {
-    return this.getLockedConnections().map((c) => c.remotePeer);
+  public async getPeers(params: GetPeersParams): Promise<PeerId[]> {
+    const connectedPeers = await this.connectionManager.getConnectedPeers();
+    let results: Peer[] = [];
+
+    for (const peer of connectedPeers) {
+      const hasProtocol = this.hasPeerProtocol(peer, params.protocol);
+      const hasSamePubsub = await this.connectionManager.isPeerOnPubsubTopic(
+        peer.id,
+        params.pubsubTopic
+      );
+      const isPeerAvailableForUse = this.isPeerAvailableForUse(peer.id);
+
+      if (hasProtocol && hasSamePubsub && isPeerAvailableForUse) {
+        results.push(peer);
+      }
+    }
+
+    const lockedPeers = results.filter((p) => this.isPeerLocked(p.id));
+
+    if (lockedPeers.length >= this.numPeersToUse) {
+      return lockedPeers.slice(0, this.numPeersToUse).map((p) => p.id);
+    }
+
+    const notLockedPeers = results.filter((p) => !this.isPeerLocked(p.id));
+
+    results = [...lockedPeers, ...notLockedPeers]
+      .slice(0, this.numPeersToUse)
+      .map((p) => {
+        this.lockPeer(p.id);
+        return p;
+      });
+
+    return results.map((p) => p.id);
   }
 
-  public requestRenew(peerId: PeerId | string): PeerId | undefined {
-    const lockedConnections = this.getLockedConnections();
-    const neededPeers = this.numPeersToUse - lockedConnections.length;
+  public async renewPeer(id: PeerId, params: GetPeersParams): Promise<void> {
+    const connectedPeers = await this.connectionManager.getConnectedPeers();
+    const renewedPeer = connectedPeers.find((p) => p.id.equals(id));
 
-    if (neededPeers === 0) {
+    if (!renewedPeer) {
+      log.warn(`Cannot renew peer:${id}, no connection to the peer.`);
       return;
     }
 
-    const connections = this.getUnlockedConnections()
-      .filter((c) => !c.remotePeer.equals(peerId))
-      .slice(0, neededPeers)
-      .map((c) => this.lockConnection(c))
-      .map((c) => c.remotePeer);
+    this.unlockPeer(renewedPeer.id);
+    await this.getPeers(params);
+  }
+
+  public async isPeerOnPubsub(
+    id: PeerId,
+    pubsubTopic: string
+  ): Promise<boolean> {
+    return this.connectionManager.isPeerOnPubsubTopic(id, pubsubTopic);
+  }
 
-    const newPeerId = connections[0];
+  /**
+   * Used to notify about new Filter peer available.
+   */
+  private async onConnected(event: CustomEvent<PeerId>): Promise<void> {
```

Comment:
The logic in this function and `onDisconnect` should not be specific to Filter, it should be generalized to allow checking if a peer supports any specified protocol. Each protocol would then be responsible for calling peer manager to add their own connection/disconnection callbacks. e.g. instead of just calling `start()`, filter would call something like `listen([Protocol.Filter])`

Reply:
I agree but this was intentional choice:
- we need such thing only for `Filter`, no other protocol uses it or will use;
- if another protocol needs it - another event `lightpush:connect` can be added;